policies: fix send deadlock in 'PolicyProcess'

lhjt opened the pull request:
## Details

This refactor updates the policy engine and evaluator components to use a multiprocessing Queue instead of a Connection for inter-process communication.

The reason for this is if the payload size for a `send()` is too large, it will block. This caused SAML responses with large payloads to hang forever.

Fixes #10801.
Closes #20120

---

## Checklist

-   [x] Local tests pass (`ak test authentik/`)
-   [x] The code has been formatted (`make lint-fix`)

lhjt committed:
fix: replace `Connection` with `Queue` in `PolicyProcess` handling

This refactor updates the policy engine and evaluator components to use a multiprocessing Queue instead of a Connection for inter-process communication.

The reason for this is if the payload size for a `send()` is too large, it will block. This caused SAML responses with large payloads to hang forever.

lhjt committed:
fix(policies): use non-blocking put for result_queue

- Use `put_nowait()` instead of `put()` to avoid blocking

rissson commented:
I've started the tests for this, you should be able to see there if anything is failing.

lhjt committed:
chore: update queue import root

lhjt committed:
fix: grab default multiprocessing ctx

lhjt commented:
I have deployed this locally and tested the SAML issue outlined in #10801 (which was the same issue I ran into). I can confirm it is working as expected and it is now ready to review and merge.

rissson committed:
remove context

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson committed:
create only one queue for all iterations

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson committed:
put_nowait

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson committed:
qsize weird stuff

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson committed:
Revert "qsize weird stuff"

This reverts commit 191e17940becd16b420210bd6ceda3707e04d878.

rissson committed:
Revert "create only one queue for all iterations"

This reverts commit 45dd2abdc2859ac8fe81a36b5969623416e8318a.

rissson committed:
none check

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson committed:
re-add get_context

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson commented on authentik/policies/engine.py:
```diff
@@ -130,24 +130,26 @@ def build(self) -> "PolicyEngine":
                 # Only call .recv() if no result is saved, otherwise we just deadlock here
                 if not proc_info.result:
-                    proc_info.result = proc_info.connection.recv()
+                    proc_info.result = proc_info.result_queue.get()
```
Having played with this a bit, I think that this call might block forever if a task times out.

lhjt replied:
Yes though it's possible for me to change this to also have its own timeout. Shall it just raise an exception in this case? And how long should the timeout be if so?

lhjt replied:
I've gone ahead and made changes to this. Please let me know if the exception raised is appropriate for this circumstance.

rissson reviewed (changes requested):
Thank you for the contribution! I've played around with this a bit, and it definitely feels like using a queue is the correct approach.

However, I'm not keen about the following things:
- creating a queue for each fork seems overkill, one queue for everything should do the trick. However, due to sync issues, retrieving the results is not as easy as it seems.
- using multiprocessing.Queue for non-forking tasks is also overkill, we should just use the standard queue.Queue in that case

There are multiple improvements to do on this, like #14130, so don't treat this like an actual review, but more as notes for myself and others.

lhjt committed:
fix(policy): account for queue get timeout

lhjt commented:
Hey @rissson would you be able to run https://github.com/goauthentik/authentik/pull/14133#issuecomment-2813399091 on this PR as well? It'll make testing it in my setup a lot faster since I can't build `amd64` images on my arm mac. Thanks 🙂

lhjt committed:
feat(engine): introduce `PolicyQueueCoordinator`

- Allows the use of a single result_queue for all results.

lhjt committed:
fix(engine): use multiprocessing context

lhjt committed:
fix(policies): ensure thread-safe access to result map with locking

Added a threading lock to the PolicyQueueCoordinator to ensure
thread-safe operations when accessing and modifying the `_result_map`.

This change prevents potential race conditions by protecting `_result_map`
access with `_map_lock`.

lhjt committed:
refactor(policies): set PolicyProcessInfo result

Fixed assignment of the result in `PolicyEngine` to ensure the `PolicyProcessInfo.result` attribute is populated correctly.

lhjt committed:
fix(engine): change `PolicyProcess` constructor signature

BeryJu committed:
Merge branch 'main' into fix/deadlock

Signed-off-by: Jens Langhammer <jens@goauthentik.io>

# Conflicts:
#	authentik/policies/engine.py
#	authentik/policies/process.py

rissson committed:
Merge branch 'main' into fix/deadlock

rissson committed:
Merge branch 'main' into fix/deadlock

rissson committed:
fix tests?

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson committed:
Merge branch 'main' into fix/deadlock
